### PR TITLE
ks-444: integrate keystone contract deployment

### PR DESCRIPTION
**Pull request by krehermann**
[KS-444](https://smartcontract-it.atlassian.net/browse/KS-444)

This add keystone specific implementation for programmatically deploying and configuration contracts.

The primary new concept is CLO integration

 - keystone uses CLO to track nops, not the job distributor. the framework expects JD and this is an adapter between the two
 - The primary use case is to consume the data model of CLO (extended by [KS-453](https://smartcontract-it.atlassian.net/browse/KS-453) ) to construct don nodesets discuss above.
 - to implement the adapter in a reasonable timeframe, i had to copy the generated data model from clo into this repo. cleanup tracked in [KS-455](https://smartcontract-it.atlassian.net/browse/KS-455)
 

 
 

[KS-444]: https://smartcontract-it.atlassian.net/browse/KS-444?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[KS-453]: https://smartcontract-it.atlassian.net/browse/KS-453?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[KS-455]: https://smartcontract-it.atlassian.net/browse/KS-455?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: integration-tests/deployment/clo/offchain_client_impl.go
Comment (krehermann) [inline integration-tests/deployment/clo/offchain_client_impl.go:1]: forked from clo repo until KS-455 can be tackled https://github.com/smartcontractkit/feeds-manager/tree/develop/api/models

File: integration-tests/deployment/clo/testdata/chain_writer_nodes.json
Comment (krehermann) [inline integration-tests/deployment/clo/testdata/chain_writer_nodes.json:1]: testdata is just the output of `fmscli get nodeOperators`.  https://github.com/smartcontractkit/feeds-manager/tree/develop?tab=readme-ov-file#chainlink-orchestrator-api-client

this kind of data will be input to the testnet deployment

File: integration-tests/deployment/keystone/deploy_test.go
Comment (krehermann) [inline integration-tests/deployment/keystone/deploy_test.go]: this test mimic what will happen for a real set of testnets.

the difference is hidden in `NewDonEnvWithMemoryChains`, which allocates a simulated backend for each chain that the node supports.

on testnet, we will inject real chains.

File: integration-tests/deployment/keystone/deploy.go
Comment (krehermann) [inline integration-tests/deployment/keystone/deploy.go]: this is the primary function to drive the deployment and configure the register.

i want to come back and simplify the code into bit size pieces, but it works as is with multiple tests and need to move forward https://smartcontract-it.atlassian.net/browse/KS-456

File: integration-tests/deployment/clo/don_nodeset.go
Comment (bolekk) [inline integration-tests/deployment/clo/don_nodeset.go]: Did you mean to compose all filters? In that case shouldn't you be using the output of previous iteration instead of nonBootstrapNops every time?

File: integration-tests/deployment/clo/models/models.go
Comment (bolekk) [inline integration-tests/deployment/clo/models/models.go:10]: Where is the source that models_gen.go was generated from? Is models.go just an addition with some handy constants?

File: integration-tests/deployment/keystone/capability_definitions.go
Comment (bolekk) [inline integration-tests/deployment/keystone/capability_definitions.go]: The whole file, it seems...? Or just the stuff down below?

File: integration-tests/deployment/keystone/deploy.go
Comment (bolekk) [inline integration-tests/deployment/keystone/deploy.go]: I don't understand - why would we have duplicates? Is this to make the tool idempotent? But we're deploying a new Registry contract on every call...

File: integration-tests/deployment/keystone/forwarder_deployer.go
Comment (bolekk) [inline integration-tests/deployment/keystone/forwarder_deployer.go:44]: Where does configuring the contract happen?

File: integration-tests/deployment/keystone/ocr3_deployer.go
Comment (bolekk) [inline integration-tests/deployment/keystone/ocr3_deployer.go:44]: Where does configuring OCR happen?

File: integration-tests/deployment/environment.go
Comment (bolekk) [inline integration-tests/deployment/environment.go]: Would you mind explaining the concept of an Environment please? And why it is important to "group nodesets into a don and compose those dons into an Environment"?

Comment (krehermann) [reply integration-tests/deployment/environment.go]: the original concept inspired by CCIP was dev vs testnet vs prod.

an environment is composed of nodes, contracts, and configuration that couple them

however the implementation that they have assumes that all the nodes in a given environment host all capabilities.

this doesn't work for keystone, and one why to accomodate our need is by logically partitioning the total set of nodes into subsets that host given sets of capabilities.

so, i leveraged the existing idea of an Env and extended it to support nodesets that host different capabilities. i modeled each of these nodesets as an Environment, although it a bit of an abuse of the original idea. but that really isn't exposed. rather, the thing that is exposed is a MultiDonEnvironment, which, again aims to group nodesets into dons that host different capabilities.

File: integration-tests/deployment/clo/don_nodeset.go
Comment (krehermann) [reply integration-tests/deployment/clo/don_nodeset.go]: no, this is partitioning nodes into dons.

add some comments and renamed the func, hopefully clearer

File: integration-tests/deployment/clo/models/models.go
Comment (krehermann) [reply integration-tests/deployment/clo/models/models.go:10]: updated comment with ptr to source

File: integration-tests/deployment/environment.go
Comment (krehermann) [reply integration-tests/deployment/environment.go]: add code comment hopefully captures the idea

File: integration-tests/deployment/keystone/forwarder_deployer.go
Comment (krehermann) [reply integration-tests/deployment/keystone/forwarder_deployer.go:44]: thanks for finding this mistake. see deploy.go starting at line 163

File: integration-tests/deployment/keystone/ocr3_deployer.go
Comment (krehermann) [reply integration-tests/deployment/keystone/ocr3_deployer.go:44]: same as above: thanks and see deploy.go ~ line 163

File: integration-tests/deployment/keystone/README.md
Comment (vyzaldysanchez) [inline integration-tests/deployment/keystone/README.md:88]: `./bin/fmscli` refers to the `feeds-manager` directory I guess?

File: integration-tests/deployment/address_book.go
Comment (krehermann) [inline integration-tests/deployment/address_book.go]: @AnieeG @connorwstein  i needed to change this. i think it's safe since it's normal for merge to overwrite and existing addr. problem was uncovered if i wanted to re-run a migration in `chainlink-deployments`

File: integration-tests/deployment/multiclient.go
Comment (krehermann) [inline integration-tests/deployment/multiclient.go:100]: @AnieeG @connorwstein here is a parallel wait mined that i used in the Confirm of `chainlink-deployments`. working well.

File: core/scripts/keystone/src/88_gen_ocr3_config.go
Comment (krehermann) [inline core/scripts/keystone/src/88_gen_ocr3_config.go:21]: @archseer this is refactored in the integration-tests. 

want to refactor the rest of these script to use common code from there

File: integration-tests/deployment/multiclient.go
Comment (AnieeG) [inline integration-tests/deployment/multiclient.go:111]: how about logging the failure as well. if we get len(mc.clients) no of failure we can return early with error instead of waiting for timeout

**Review by cedric-cordenier — DISMISSED** [review]

File: core/scripts/go.mod
Comment (archseer) [inline core/scripts/go.mod:8]: Should remove this before the merge

File: core/scripts/go.mod
Comment (jmank88) [inline core/scripts/go.mod:28]: What causes this dependency? It seems backwards from what I would expect. Would another module for them both to share be ideal?

File: integration-tests/deployment/address_book.go
Comment (connorwstein) [reply integration-tests/deployment/address_book.go]: The intention with the strictness here was to avoid silently failing in this scenario: someone writes a buggy changeset which was supposed to deploy a new address, but it returns an existing address instead. In that case the changeset will succeed but the user is unaware of the issue. To iterate on changesets you can:
1. Run changeset A, generates artifacts/A...
2. Inspect artifacts, discover an issue. Remove artifacts/A...
3. Repeat 1-2 until satisfied. 
4. Promote artifacts via CLI/CI.

File: integration-tests/deployment/keystone/capability_registry_deployer.go
Comment (connorwstein) [inline integration-tests/deployment/keystone/capability_registry_deployer.go]: is this address irrelevant?

File: integration-tests/deployment/keystone/capability_registry_deployer.go
Comment (connorwstein) [inline integration-tests/deployment/keystone/capability_registry_deployer.go]: strictly for estimating deployment gas? Maybe estimateDeployGas

File: integration-tests/deployment/keystone/contract_set.go
Comment (connorwstein) [inline integration-tests/deployment/keystone/contract_set.go]: we moved to this pattern instead: 
```
// Caller knows ab will be mutated. 
func DeployA(ab addressBook, ..) error {
	// do deploy 
	if err := ab.Save(..); err != nil {
		return err
	}
}

func DeployAB(ab addressBook, ..) error {
	if err := DeployA(ab,...) err != nil {
		return error
	}
	// deploy B and save 
}

// Only a test or top level changeset instantiates the address book. 
func TestAB(t *testing.T) {
	ab := deployment.NewAddressBook()
	err := DeployAB(ab)
	if err != nil {
		// If there's an error we can still check the ab to see which contracts were deployed if desired.
	} {
		// no error the ab should have both A and B as expected, we can load contract wrappers using it etc. 
	}
}
```
Might be simpler, you avoid merging generally

File: integration-tests/deployment/keystone/deploy.go
Comment (connorwstein) [inline integration-tests/deployment/keystone/deploy.go:120]: Intention here is you'd put this in a changeset package like https://github.com/smartcontractkit/chainlink/blob/9efabb039f78c1890a4a3749fb4986d6b7947733/integration-tests/deployment/ccip/changeset/cap_reg.go#L11 so its clear what you want to export to higher envs.

File: integration-tests/deployment/keystone/deploy.go
Comment (connorwstein) [inline integration-tests/deployment/keystone/deploy.go]: should be able to use shared helpers here https://github.com/smartcontractkit/chainlink/blob/cd7c32c0b7401858966ac20de4a3fca74673ab71/integration-tests/deployment/helpers.go#L66

File: integration-tests/deployment/keystone/deploy.go
Comment (connorwstein) [inline integration-tests/deployment/keystone/deploy.go:603]: I think you'll want to separate calldata generation from the actual call here, assuming you want the registry to be owned by MCMS. See for example here https://github.com/smartcontractkit/chainlink/blob/9efabb039f78c1890a4a3749fb4986d6b7947733/integration-tests/deployment/ccip/add_chain.go#L131 where BuildAddDONArgs is used for in an MCMS proposal context as well as in a non-MCMS (deployment) context https://github.com/smartcontractkit/chainlink/blob/9efabb039f78c1890a4a3749fb4986d6b7947733/integration-tests/deployment/ccip/deploy_home_chain.go#L382

File: integration-tests/deployment/keystone/deploy.go
Comment (connorwstein) [inline integration-tests/deployment/keystone/deploy.go]: nit: I'd probably invert this to limit the nesting and improve readability: 
```
if !strings.Contains(..) {
	return ..
}
...
```

File: integration-tests/deployment/keystone/deploy.go
Comment (connorwstein) [inline integration-tests/deployment/keystone/deploy.go:656]: maybe a way to use/extend sorted methods on https://github.com/smartcontractkit/chainlink/blob/5c51b27bb1350645d08afab2e98e56ba84ae5c28/integration-tests/deployment/environment.go#L128?

File: integration-tests/deployment/keystone/deploy_test.go
Comment (connorwstein) [inline integration-tests/deployment/keystone/deploy_test.go]: why the single subtest here vs just top level?

File: integration-tests/deployment/keystone/ocr3_deployer.go
Comment (connorwstein) [inline integration-tests/deployment/keystone/ocr3_deployer.go:21]: you could probably follow the same pattern we have with Go generics here https://github.com/smartcontractkit/chainlink/blob/9efabb039f78c1890a4a3749fb4986d6b7947733/integration-tests/deployment/ccip/deploy.go#L92 to avoid unique deploy methods per contract

File: integration-tests/deployment/keystone/ocr3config.go
Comment (connorwstein) [inline integration-tests/deployment/keystone/ocr3config.go:103]: we can come back to it, but there should  be a way to can share OCR3/Capability registry management code in for example some deployment/capreg package.

File: integration-tests/deployment/keystone/state.go
Comment (connorwstein) [inline integration-tests/deployment/keystone/state.go:23]: for a follow up, when you want to generate "RDD style" data you can follow the "view" package pattern we did here https://github.com/smartcontractkit/chainlink/tree/815b2c80ff6b0dff940497f7734aff3662e7b4f9/integration-tests/deployment/ccip/view  (versioned, json marshallable structs) generated from this ContractSet

File: integration-tests/deployment/keystone/types.go
Comment (connorwstein) [inline integration-tests/deployment/keystone/types.go:54]: you sure you can't go straight from OffchainClient -> Node https://github.com/smartcontractkit/chainlink/blob/5c51b27bb1350645d08afab2e98e56ba84ae5c28/integration-tests/deployment/environment.go#L173 -> Capability registration? We did that here https://github.com/smartcontractkit/chainlink/blob/9efabb039f78c1890a4a3749fb4986d6b7947733/integration-tests/deployment/ccip/deploy_home_chain.go#L200

File: integration-tests/deployment/memory/chain.go
Comment (connorwstein) [inline integration-tests/deployment/memory/chain.go:77]: why specific IDs?

File: integration-tests/deployment/multiclient.go
Comment (connorwstein) [inline integration-tests/deployment/multiclient.go]: probably a wait group would be cleaner here? then you can just wg.Wait() until they all return

File: integration-tests/deployment/multiclient.go
Comment (connorwstein) [inline integration-tests/deployment/multiclient.go]: to remove?

File: integration-tests/deployment/multiclient.go
Comment (connorwstein) [inline integration-tests/deployment/multiclient.go]: to remove

File: integration-tests/deployment/multiclient.go
Comment (connorwstein) [inline integration-tests/deployment/multiclient.go]: ditto

**Review by connorwstein — CHANGES_REQUESTED** [review]

File: integration-tests/deployment/keystone/capability_registry_deployer.go
Comment (krehermann) [reply integration-tests/deployment/keystone/capability_registry_deployer.go]: yes, removed.

File: integration-tests/deployment/keystone/capability_registry_deployer.go
Comment (krehermann) [reply integration-tests/deployment/keystone/capability_registry_deployer.go]: done

File: integration-tests/deployment/keystone/contract_set.go
Comment (krehermann) [reply integration-tests/deployment/keystone/contract_set.go]: yeah, i think the existing pattern that i have led to me wanting to change the impl of Merge.

File: integration-tests/deployment/keystone/deploy.go
Comment (krehermann) [reply integration-tests/deployment/keystone/deploy.go:120]: added

File: integration-tests/deployment/memory/chain.go
Comment (krehermann) [reply integration-tests/deployment/memory/chain.go:77]: it made it easier to test keystone where not all chains are equal. specifically that the capability registry and ocr 3 contracts are on the expected chain.

this may be cruft that can be ripped out, but i'm hesitant to do that now given that it's working

File: integration-tests/deployment/multiclient.go
Comment (krehermann) [reply integration-tests/deployment/multiclient.go]: sleep was leftover hacking in the early stage. no need to wait; removed.

File: integration-tests/deployment/keystone/ocr3_deployer.go
Comment (krehermann) [reply integration-tests/deployment/keystone/ocr3_deployer.go:21]: yeah. i thought unique deploy funcs are easier to understand and simpler to start, if more verbose. may revisit

File: integration-tests/deployment/keystone/deploy_test.go
Comment (krehermann) [reply integration-tests/deployment/keystone/deploy_test.go]: fixed; left over cruft

File: integration-tests/deployment/keystone/deploy.go
Comment (krehermann) [reply integration-tests/deployment/keystone/deploy.go:603]: will do this in a followup.

as discussed offline, the initial deployment and configuration will be done by the deployer key. changesets after that will need to go thru MCMS and will have to proposed proposals

File: integration-tests/deployment/keystone/types.go
Comment (krehermann) [reply integration-tests/deployment/keystone/types.go:54]: added a ticket and todo that can be address when we fully migrate to JD.

this struct is an internal helper, so there is no risk to refactoring it later.

File: core/scripts/go.mod
Comment (krehermann) [reply core/scripts/go.mod:8]: i think it's ok since we are using the same trick in the line above.

in any case, i can do pin to a specific version after this is merged, but before it's merged the version is not on develop and errors some ci check

File: core/scripts/go.mod
Comment (krehermann) [reply core/scripts/go.mod:28]: yes, we can refactor to share as a followup.

the keystone scripts will be rm'ed/absorbed in the the integration tests deployment. this depends is a mid term solution so that they don't diverge.

i choose this way rather than importing scripts 1. b/c scripts will be removed over time 2. there is another repo, `chainlink-deploments` that uses code in this pr to deploy to testnet/mainnet. it already heavily depends on integration-tests and so it was simpler to keep that pattern rather than split-brain between scripts & integration-tests

File: integration-tests/deployment/keystone/changeset/0001_deploy_registry.go
Comment (connorwstein) [inline integration-tests/deployment/keystone/changeset/0001_deploy_registry.go:11]: note - we found that leaving these unordered is actually easier. Since you will have multiple higher environments (staging/testnet/mainnet) where you may want to run them in different orders/skip some etc. Can take a look at the CCIP dir for naming there.

**Review by connorwstein — DISMISSED** [review]

**Review by archseer — DISMISSED** [review]

**Comment by archseer** [comment]
We probably want to add the keystone team as codeowners for `integration-tests/deployment/keystone`

https://github.com/smartcontractkit/chainlink/blob/002296d9dbbcb356d6f66d5e274b18bfb16681be/.github/CODEOWNERS#L108

**Review by cedric-cordenier — DISMISSED** [review]

**Review by cedric-cordenier — APPROVED** [review]

**Review by connorwstein — APPROVED** [review]

**Pull request merged**
